### DISCUSSION TITLE
Is in-memory mode suitable for production?

### DISCUSSION THREAD

**Discussion by jerve** [phax/phase4#279] [Q&A] [answered]
I was wondering if the in-memory mode of phase4 is suitable for production.
I detected it usages storage for 4 different types of domain objects:

- Processing modes
- Duplicates
- Profiles
- MPC

It's not clear to me what would be the side effect if those objects where ephemeral (are gone when the service is restarted).
Some insights would be great.

**Comment by phax** [comment, marked as answer]
Hi @jerve, basically it depends on the use case your are planning to use it in.
If you are thinking about using phase4 in the context of Peppol, than here are my thoughts:
* PMode manager: not needed, because everything is dynamic via the SMP lookup anyway - so in-memory is production fine
* Profile manager: the profile is only registered on a code level via SPI callbacks - so in-memory is production fine
* MPC manager: not needed because, MPC are primarily needed for Pull Requests which are not used in Peppol - so in-memory is production fine
* Duplicate manager: this handles retries on the AS4 level. The current implementation checks for duplicates in the last 10 minutes (customizable). I have never heard of production issues with that, however I can see the need to provide a more robust implementation for that. Building e.g. a MongoDB backend is pretty straight forward - using an SQL backend would be a little more challenging.

hth

**Reply by jerve** [reply]
Hi phax, thanks for your insights. I am indeed planning to use it for Peppol, so in-memory should be good for now. Thanks!
